plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/embeddings/TypeEmbedding.kt

```diff
@@ -251,5 +261,49 @@ data class ClassTypeEmbedding(val className: ScopedKotlinName) : TypeEmbedding {
     override fun accessInvariants(v: Exp): List<Exp> =
         flatMapUniqueFields { _, field -> field.accessInvariantsForParameter(v) }
 
+    // Note: this function will replace accessInvariants when nested unfold will be implemented
+    override fun predicateAccessInvariants(v: Exp): List<Exp> = listOf(Exp.PredicateAccess(name, listOf(v)))
+
     override fun provenInvariants(v: Exp) = listOf(subtypeInvariant(v))
+
+    private val thisEmbedding = VariableEmbedding(ThisReceiverName, this)
+    val classPredicate: Predicate
+        get() {
+            val accessReadableFields = fields.values
+                .filter { it.accessPolicy == AccessPolicy.ALWAYS_READABLE }
+                .map { FieldAccess(thisEmbedding, it).getAccessPredicate(PermExp.WildcardPerm()) }
+
+            val accessSpecialFields = fields.values
+                .filter { it.accessPolicy == AccessPolicy.ALWAYS_WRITEABLE }
+                .map { FieldAccess(thisEmbedding, it).getAccessPredicate(PermExp.FullPerm()) }
+
+            val classFieldPredicates = fields.values
+                .filter { it.accessPolicy == AccessPolicy.ALWAYS_READABLE }
+                .map { it.type.predicateAccessInvariants(Exp.FieldAccess(Exp.LocalVar(ThisReceiverName, Type.Ref), it.toViper())) }
+                .flatten()
+
+            val superTypesPredicates = superTypes
+                .filterIsInstance<ClassTypeEmbedding>()
+                .map { Exp.PredicateAccess(it.name, listOf(thisEmbedding.toViper())) }
+
+            val conditions = accessReadableFields + accessSpecialFields + specialConditions + classFieldPredicates + superTypesPredicates
+            val body = if (conditions.isNotEmpty()) {
+                conditions
+                    .drop(1)
+                    .fold(conditions.first()) { acc, expEmbedding -> Exp.And(acc, expEmbedding) }
+            } else {
+                Exp.BoolLit(true)
+            }
+            return Predicate(name, listOf(thisEmbedding.toLocalVarDecl()), body)
+        }
+
+    private val specialConditions: List<Exp>
```

Comment:
Could you open an issue on space for making a single extension point for this kind of thing?  I think it's good to be able to add ad-hoc stuff to functions and types, but the fact it's happening in so many places in the code can become a problem.

(Consider also whether putting this with the others and making `thisEmbedding` a parameter wouldn't be clearer.)

Reply:
Do you mean an issue to lighten classes in `TypeEmbedding`?

Reply:
An issue to make sure that the special cases we have in our code (e.g. for stdlib functions) should eventually all end up in one file/module, rather than being spread out.

Reply:
ok, anyway after some modifications, this function no longer exists